## ------------------------------------------------------------------------------------ ##
## Preparation - set software paths
## ------------------------------------------------------------------------------------ ##
## Path to R binary and library
R := R_LIBS=/home/Shared/Rlib/release-3.5-lib/ /usr/local/R/R-3.4.0/bin/R CMD BATCH --no-restore --no-save

## Path to Salmon binary
salmon := /home/charlotte/software/Salmon-0.8.2_linux_x86_64/bin/salmon

## Path to TrimGalore! and cutadapt, if adapter trimming is performed
trimgalore := /home/charlotte/software/trim_galore_v0.4.4/trim_galore
cutadapt := /home/charlotte/.local/bin/cutadapt

## Path to STAR
STAR := /home/Shared_penticton/software/STAR/source/STAR

## Path to samtools binary
samtools := /usr/local/bin/samtools

## Path to MultiQC 

## ------------------------------------------------------------------------------------ ##
## Preparation - set paths to reference files
## ------------------------------------------------------------------------------------ ##
## Path to existing genome fasta file
genome := 

## Path to STAR genome index
STARindex := $(refdir)/STARIndex/Danio_rerio.GRCz10.dna.toplevel_Overhang149

## Path to existing cDNA and ncRNA fasta files
cdna := 
ncrna := 

## Path to merged cDNA and ncRNA fasta file (will be generated by the script)
txome := 

## Path to Salmon index
salmonindex := 

## Path to rds file where the tx2gene mapping will be stored
tx2gene := 

## ------------------------------------------------------------------------------------ ##
## Preparation - list files to process
## ------------------------------------------------------------------------------------ ##
## List sample IDs (should correspond to the unique parts of FASTQ file names
samples := 

.PHONY: all

## ------------------------------------------------------------------------------------ ##
## Target definition
## ------------------------------------------------------------------------------------ ##
## Run all analyses
all: multiqc/multiqc_report.html

## List all the packages that were used by the R analyses
listpackages:
	$(R) scripts/list_packages.R Rout/list_packages.Rout

## ------------------------------------------------------------------------------------ ##
## Reference preparation
## ------------------------------------------------------------------------------------ ##
## Salmon - generate index from merged cDNA and ncRNA files
$(salmonindex)/hash.bin: $(txome)
	$(salmon) index -t $< -k 31 -i $(@D) --type quasi

## Generate tx2gene mapping
tx2gene: $(txome)
	$(R) "--args transcriptfasta='$(txome)' outrds='$@'" Rscripts/generate_tx2gene.R Rout/generate_tx2gene.Rout

## ------------------------------------------------------------------------------------ ##
## Adapter trimming